snapcraft/internal/build_providers/_snap.py

```diff
@@ -0,0 +1,347 @@
+            self._set_data()
+        return self.__revision
+
+    def get_install_cmd(self) -> List[str]:
+        if self.__install_cmd is None:
+            self._set_data()
+        return self.__install_cmd
+
+
+_STORE_ASSERTION = [
+    "account-key",
+    "public-key-sha3-384=BWDEoaqyr25nF5SNCvEv2v7QnM9QsfCc0PBMYD_i2NGSQ32EF2d4D0hqUel3m8ul",
+]
+
+
+def _load_registry(registry_filepath: Optional[str]) -> Dict[str, List[Any]]:
+    if registry_filepath is None or not os.path.exists(registry_filepath):
+        return dict()
+
+    with open(registry_filepath) as registry_file:
+        return yaml.load(registry_file)
+
+
+def _save_registry(
+    registry_data: Dict[str, List[Any]], registry_filepath: Optional[str]
+) -> None:
+    if registry_filepath is None:
+        return
+
+    dirpath = os.path.dirname(registry_filepath)
+    if dirpath:
+        os.makedirs(dirpath, exist_ok=True)
+
+    with open(registry_filepath, "w") as registry_file:
+        yaml.dump(registry_data, stream=registry_file)
+
+
+class SnapInjector:
+    """Handle the process of adding snaps into the build environment.
+
+    The specific knowledge of the build environment where these snaps will
+    be injected is not required, instead runnables to execute the required
+    operations against the build environment are provided upon initialization.
+
+    The snaps to install or refresh in the build environment are added by calling
+    add and finally applied by calling apply. If an operation is required, the
+    revision that eventually made it into the environment is recorded in the
+    registry.
+    """
+
+    def __init__(
+        self,
+        *,
+        snap_dir: str,
+        registry_filepath: str,
+        runner: Callable[..., None],
+        snap_dir_mounter: Callable[[], None],
+        snap_dir_unmounter: Callable[[], None],
+        file_pusher: Callable[..., None]
+    ) -> None:
+        """
+        Initialize a SnapInjector instance.
+
+        :param str snap_dir: directory where snaps from the host live for the cases
+                             where injection of the snap into the build environment
+                             is possible.
+        :param str registry_filepath: path to where recordings of previusly installed
+                                      revisions of a snap can be queried and recorded.
+        :param runner: a callable which can run commands in the build environment.
+        :param snap_dir_mounter: a callable which can mount the directory where snaps live
+                                 on the host into the build environment. This callable
+                                 must mount into snap_dir.
+        :param snap_dir_unmounter: a callable which can unmount snap_dir from the environment.
+        :param file_pusher: a callable that can push file from the host into the build
+                            environment.
+        """
+
+        self._snaps = []  # type: List["_SnapManager"]
```

Comment:
Does this type need to be quoted? It's defined earlier in this file.

Reply:
only because I moved things around, probably not